server/src/main/java/org/elasticsearch/action/search/FetchSearchPhase.java

```diff
@@ -158,30 +158,31 @@ private void innerRunFetch(ScoreDoc[] scoreDocs, int numShards, SearchPhaseContr
         );
         for (int i = 0; i < docIdsToLoad.length; i++) {
             List<Integer> entry = docIdsToLoad[i];
-            RankDocShardInfo rankDocs = rankDocsPerShard == null || rankDocsPerShard.get(i).isEmpty()
-                ? null
-                : new RankDocShardInfo(rankDocsPerShard.get(i));
-            SearchPhaseResult shardPhaseResult = searchPhaseShardResults.get(i);
             if (entry == null) { // no results for this shard ID
-                if (shardPhaseResult != null) {
-                    // if we got some hits from this shard we have to release the context there
-                    // we do this as we go since it will free up resources and passing on the request on the
-                    // transport layer is cheap.
-                    releaseIrrelevantSearchContext(shardPhaseResult, context);
-                    progressListener.notifyFetchResult(i);
-                }
+                // if we got some hits from this shard we have to release the context
+                // we do this below after sending out the fetch requests relevant to the search to give priority to those requests
+                // that contribute to the final search response
                 // in any case we count down this result since we don't talk to this shard anymore
                 counter.countDown();
```

Comment:
Should the count down happen **after** sending the release context?

Reply:
I figured no :) We don't wait for the response to the request, why wait for sending it out?
Note that we are not super efficient here and this can take O(100ms) time for thousands of shards, no need for the REST layer to wait for that long (and at this scale almost more importantly, us to waste heap while we loop over shipping them out)